### ISSUE TITLE
IllegalTokenTextCheck produces false positives for octal and escaped backslash sequences in string literals

### ISSUE THREAD

**Issue by Atharv3221** [checkstyle/checkstyle#19954]
Issue for Google style is described at https://github.com/checkstyle/checkstyle/issues/19548

I have read check documentation: https://checkstyle.org/checks/coding/illegaltokentext.html#IllegalTokenText
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
atharv@atharv-IdeaPad-3-15IIL05:~/test$ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="IllegalTokenText">
      <property name="tokens" value="STRING_LITERAL"/>
      <property name="format" value="\\11"/>
    </module>
  </module>
</module>

```

```
atharv@atharv-IdeaPad-3-15IIL05:~/test$ cat Test2.java
public class Test2 {

    public void openjdkCase() {
        String a = "\117"; // no violation expected

        String b = "\118"; // violation
    }

    public void guavaCase() {
        String c = "\\11"; // no violation expected

        String d = "\11"; // violation
    }
}
```

```
atharv@atharv-IdeaPad-3-15IIL05:~/test$ java -jar checkstyle-13.4.2-all.jar -c config.xml Test2.java
Starting audit...
[ERROR] /home/atharv/test/Test2.java:4:20: Token text matches the illegal pattern '\\11'. [IllegalTokenText]
[ERROR] /home/atharv/test/Test2.java:6:20: Token text matches the illegal pattern '\\11'. [IllegalTokenText]
[ERROR] /home/atharv/test/Test2.java:10:20: Token text matches the illegal pattern '\\11'. [IllegalTokenText]
[ERROR] /home/atharv/test/Test2.java:12:20: Token text matches the illegal pattern '\\11'. [IllegalTokenText]
Audit done.
Checkstyle ends with 4 errors.
```

Expected:

```
Starting audit...
[ERROR] /home/atharv/test/Test2.java:6:20: Token text matches the illegal pattern '\\11'. [IllegalTokenText]
[ERROR] /home/atharv/test/Test2.java:12:20: Token text matches the illegal pattern '\\11'. [IllegalTokenText]
Audit done.
Checkstyle ends with 2 errors.
```
Ref: https://github.com/checkstyle/checkstyle/pull/19549#issuecomment-4232176395

**Comment by Zopsss** [comment]
You have only provided the code examples, you should also explain the false positives. What is the problem and how we can solve it.

**Comment by Atharv3221** [comment]
Currently the check analyzes raw token text purely through regex substring matching (`Matcher.find()`), without understanding Java lexical escape boundaries defined in [JLS 3.3](https://docs.oracle.com/javase/specs/jls/se21/html/jls-3.html#jls-3.3) and [JLS 3.10.7](https://docs.oracle.com/javase/specs/jls/se21/html/jls-3.html#jls-3.10.7).

This creates a fundamental limitation: the regex engine only sees a character stream, while the Java lexer interprets escape sequences greedily and contextually.

For example:

* In the OpenJDK case, `\100` is a single 3-digit octal escape. The Java lexer consumes all three digits as one escape sequence. However, `Matcher.find()` still sees the raw substring `\10` inside `\100` and incorrectly reports it as a separate 2-digit octal escape.

* In the Guava case, `\\11` represents an escaped backslash followed by literal characters `11`. Lexically, the first two backslashes form one complete escape sequence (`\\`). But regex matching can still start from the second backslash and incorrectly interpret `\11` as an octal escape because it has no notion of already-consumed escape boundaries.


A complete fix would require escape-sequence-aware matching instead of plain regex substring matching.

Currently `IllegalTokenText` applies regex directly on raw token text, so matching operates purely on character substrings. This works for simple cases, but it cannot reliably distinguish whether characters belong to:

* a complete octal escape,
* a unicode escape,
* an escaped backslash,
* or ordinary literal text.

To fully avoid these false positives, the check would first need to parse token text according to Java lexical escape rules and identify valid escape-sequence boundaries before applying pattern checks.

The matching logic would need to made escape-sequence-aware.

**Comment by romani** [comment]
Here is explanation that this Check can not cover all requirements https://github.com/checkstyle/checkstyle/pull/20631#discussion_r3643308778

We need new Check to be created to smartly parse string value and identify escapes.

No changes to IllegalTokenText should be done.
